File: script/weapm/weapm_cli.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"time"
)

// ==================== 命令行参数 ====================

type CommandLineArgs struct {
	ConfigPath  string
	Env         string
	BaseURL     string
	Username    string
	Password    string
	Timeout     int
	Quiet       bool
	Command     string
	ClusterName string
	Detail      bool
	Search      bool
	SubsysID    string
	Check       string
	Limit       int
	Address     string
	Role        string
	CpuLimit    string
	MemLimit    string
	Topic       string
	BucketNames string
	BackendDomain string
	StorageDomain string
	Status      string
}

func parseArgs() *CommandLineArgs {
	args := &CommandLineArgs{}

	// 全局参数
	flag.StringVar(&args.ConfigPath, "config", "", "配置文件路径")
	flag.StringVar(&args.ConfigPath, "c", "", "配置文件路径 (简写)")
	flag.StringVar(&args.Env, "env", "", "环境名称 (dev/prod)")
	flag.StringVar(&args.Env, "e", "", "环境名称 (简写)")
	flag.StringVar(&args.BaseURL, "base-url", "", "API 基础 URL")
	flag.StringVar(&args.Username, "username", "", "用户名")
	flag.StringVar(&args.Password, "password", "", "密码")
	flag.IntVar(&args.Timeout, "timeout", 30, "请求超时时间(秒)")
	flag.BoolVar(&args.Quiet, "quiet", false, "静默模式,不输出日志")
	flag.BoolVar(&args.Quiet, "q", false, "静默模式 (简写)")

	// 集群管理参数
	flag.StringVar(&args.ClusterName, "cluster-name", "", "集群名称")
	flag.StringVar(&args.ClusterName, "n", "", "集群名称 (简写)")
	flag.BoolVar(&args.Detail, "detail", false, "显示详细信息")
	flag.BoolVar(&args.Detail, "d", false, "显示详细信息 (简写)")

	// 子系统参数
	flag.BoolVar(&args.Search, "search", false, "搜索子系统")
	flag.BoolVar(&args.Search, "s", false, "搜索子系统 (简写)")
	flag.StringVar(&args.SubsysID, "subsys-id", "", "子系统ID")
	flag.StringVar(&args.Check, "check", "", "检查子系统是否存在")
	flag.IntVar(&args.Limit, "limit", 20, "返回结果数量限制")
	flag.IntVar(&args.Limit, "l", 20, "返回结果数量限制 (简写)")

	// 节点管理参数
	flag.StringVar(&args.Address, "address", "", "节点IP地址")
	flag.StringVar(&args.Role, "role", "", "节点角色")
	flag.StringVar(&args.CpuLimit, "cpulimit", "", "CPU限制")
	flag.StringVar(&args.MemLimit, "memlimit", "", "内存限制")
	flag.StringVar(&args.Topic, "topic", "", "Topic")
	flag.StringVar(&args.BucketNames, "bucketnames", "", "存储桶名称")
	flag.StringVar(&args.BackendDomain, "backenddomain", "", "后端域")
	flag.StringVar(&args.StorageDomain, "storagedomain", "", "存储域")
	flag.StringVar(&args.Status, "status", "", "状态")

	flag.Parse()

	// 获取命令 (第一个非标志参数)
	if len(flag.Args()) > 0 {
		args.Command = flag.Args()[0]
	}

	return args
}

// ==================== 命令处理函数 ====================

func cmdDashboard(client *Client) error {
	ctx := context.Background()
	dashboard, err := client.GetDashboard(ctx)
	if err != nil {
		return err
	}

	result, _ := json.MarshalIndent(dashboard, "", "  ")
	fmt.Println(string(result))
	return nil
}

func cmdClusters(client *Client, args *CommandLineArgs) error {
	ctx := context.Background()

	if args.Detail {
		if args.ClusterName == "" {
			return fmt.Errorf("使用 --detail 时必须指定 --cluster-name")
		}

		result, err := client.GetClusterDetail(ctx, args.ClusterName)
		if err != nil {
			return err
		}

		output, _ := json.MarshalIndent(result, "", "  ")
		fmt.Println(string(output))
	} else {
		clusters, err := client.GetClusters(ctx)
		if err != nil {
			return err
		}

		output, _ := json.MarshalIndent(clusters, "", "  ")
		fmt.Println(string(output))
	}

	return nil
}

func cmdSubsystems(client *Client, args *CommandLineArgs) error {
	ctx := context.Background()

	var result interface{}
	var err error

	if args.Search {
		result, err = client.SearchSubsystems(ctx, &SearchSubsystemsRequest{
			SubsysID: &args.SubsysID,
			Limit:    args.Limit,
		})
	} else if args.Check != "" {
		result, err = client.CheckSubsystemExists(ctx, args.Check)
	} else if args.Detail != "" {
		result, err = client.GetSubsystemDetail(ctx, args.Detail)
	} else {
		result, err = client.GetSubsystems(ctx)
	}

	if err != nil {
		return err
	}

	output, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(output))
	return nil
}

func cmdAddNode(client *Client, args *CommandLineArgs) error {
	ctx := context.Background()

	node := &AddClusterNodeRequest{
		Address:       args.Address,
		Role:          args.Role,
		CpuLimit:      args.CpuLimit,
		MemLimit:      args.MemLimit,
		Topic:         args.Topic,
		BucketNames:   args.BucketNames,
		BackendDomain: args.BackendDomain,
		StorageDomain: args.StorageDomain,
		Status:        args.Status,
	}

	err := client.AddClusterNode(ctx, args.ClusterName, node)
	if err != nil {
		return err
	}

	fmt.Println(`{"code": 0, "message": "节点添加成功"}`)
	return nil
}

func cmdDeleteNode(client *Client, args *CommandLineArgs) error {
	ctx := context.Background()

	// 从 args 中获取 IP
	ip := ""
	if len(flag.Args()) > 1 {
		ip = flag.Args()[1]
	}

	if ip == "" {
		return fmt.Errorf("请指定节点IP地址")
	}

	err := client.DeleteClusterNode(ctx, ip)
	if err != nil {
		return err
	}

	fmt.Println(`{"code": 0, "message": "节点删除成功"}`)
	return nil
}

// ==================== 主函数 ====================

func main() {
	args := parseArgs()

	// 如果没有指定命令,显示帮助
	if args.Command == "" {
		fmt.Println("WEAPM-LOGSERVER API 客户端命令行工具")
		fmt.Println("\n使用方法:")
		fmt.Println("  weapm_cli <命令> [参数]")
		fmt.Println("\n可用命令:")
		fmt.Println("  dashboard    获取数据大盘信息")
		fmt.Println("  clusters     集群管理")
		fmt.Println("  subsystems   子系统管理")
		fmt.Println("  add-node     添加集群节点")
		fmt.Println("  delete-node  删除集群节点")
		fmt.Println("\n示例:")
		fmt.Println("  ./weapm_cli dashboard")
		fmt.Println("  ./weapm_cli clusters")
		fmt.Println("  ./weapm_cli clusters --detail --cluster-name LOG001")
		fmt.Println("  ./weapm_cli subsystems")
		fmt.Println("  ./weapm_cli subsystems --search --subsys-id SYS001")
		fmt.Println("  ./weapm_cli add-node --cluster-name LOG008 --address 127.0.0.2 --role write")
		fmt.Println("\n使用 --help 查看详细帮助")
		os.Exit(0)
	}

	// 配置日志
	if args.Quiet {
		log.SetOutput(os.NewFile(0, os.DevNull))
	}

	// 加载配置
	var config *Config
	var err error

	if args.ConfigPath != "" || args.Env != "" {
		config, err = LoadConfigFromYAML(args.ConfigPath, args.Env)
	} else if args.BaseURL != "" {
		// 使用命令行参数创建配置
		config = DefaultConfig(args.BaseURL)
		if args.Username != "" {
			config.Username = args.Username
		}
		if args.Password != "" {
			config.Password = args.Password
		}
		if args.Timeout != 30 {
			config.Timeout = time.Duration(args.Timeout) * time.Second
		}
	} else {
		// 默认使用配置文件
		config, err = LoadConfigFromYAML("", "")
	}

	if err != nil {
		log.Fatalf("⚠️  %v\n请先创建配置文件 config.yaml,参考 config.yaml.example", err)
	}

	// 创建客户端
	client := NewClient(config)

	// 执行命令
	var cmdErr error
	switch args.Command {
	case "dashboard":
		cmdErr = cmdDashboard(client)
	case "clusters":
		cmdErr = cmdClusters(client, args)
	case "subsystems":
		cmdErr = cmdSubsystems(client, args)
	case "add-node":
		cmdErr = cmdAddNode(client, args)
	case "delete-node":
		cmdErr = cmdDeleteNode(client, args)
	default:
		log.Fatalf("❌ 未知命令: %s", args.Command)
	}

	if cmdErr != nil {
		log.Fatalf("❌ 错误: %v", cmdErr)
	}
}

File: script/weapm/weapm_cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
WEAPM-LOGSERVER API 客户端命令行工具

基于 WEAPM-LOGSERVER REST API 的 Python 命令行客户端
支持数据大盘、集群管理、子系统运维等功能
"""

import argparse
import sys
import json
import logging
from pathlib import Path

# 导入主客户端模块
from weapm_client import WeapmClient, WeapmConfig


# 配置日志
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


# ==================== 命令处理函数 ====================

def cmd_dashboard(args, client):
    """获取数据大盘信息"""
    dashboard = client.get_dashboard()
    result = dashboard.get('result', {})
    print(json.dumps(result, indent=2, ensure_ascii=False))


def cmd_clusters(args, client):
    """获取集群列表"""
    if args.detail:
        if not args.cluster_name:
            print("错误: 使用 --detail 时必须指定 --cluster-name")
            sys.exit(1)
        result = client.get_cluster_detail(args.cluster_name)
        print(json.dumps(result.get('result', {}), indent=2, ensure_ascii=False))
    else:
        clusters = client.get_clusters()
        print(json.dumps(clusters, indent=2, ensure_ascii=False))


def cmd_subsystems(args, client):
    """获取子系统列表"""
    if args.search:
        result = client.search_subsystems(subsys_id=args.subsys_id, limit=args.limit)
        print(json.dumps(result, indent=2, ensure_ascii=False))
    elif args.check:
        result = client.check_subsystem_exists(args.check)
        print(json.dumps(result.get('result', {}), indent=2, ensure_ascii=False))
    elif args.detail:
        result = client.get_subsystem_detail(args.detail)
        print(json.dumps(result.get('result', {}), indent=2, ensure_ascii=False))
    else:
        result = client.get_subsystems()
        print(json.dumps(result, indent=2, ensure_ascii=False))


def cmd_add_node(args, client):
    """添加集群节点"""
    result = client.add_cluster_node(
        cluster_name=args.cluster_name,
        address=args.address,
        role=args.role,
        cpulimit=args.cpulimit,
        memlimit=args.memlimit,
        topic=args.topic,
        bucketnames=args.bucketnames,
        backenddomain=args.backenddomain,
        storagedomain=args.storagedomain,
        status=args.status
    )
    print(json.dumps(result, indent=2, ensure_ascii=False))


def cmd_delete_node(args, client):
    """删除集群节点"""
    result = client.delete_cluster_node(args.ip)
    print(json.dumps(result, indent=2, ensure_ascii=False))


# ==================== 参数解析 ====================

def parse_args():
    """解析命令行参数"""
    parser = argparse.ArgumentParser(
        description='WEAPM-LOGSERVER API 客户端命令行工具',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  # 获取数据大盘信息
  python weapm_cli.py dashboard

  # 获取所有集群
  python weapm_cli.py clusters

  # 获取集群详情
  python weapm_cli.py clusters --detail --cluster-name LOG001

  # 获取所有子系统
  python weapm_cli.py subsystems

  # 搜索子系统
  python weapm_cli.py subsystems --search --subsys-id SYS001

  # 检查子系统是否存在
  python weapm_cli.py subsystems --check SYS001

  # 获取子系统详情
  python weapm_cli.py subsystems --detail SYS001

  # 添加集群节点 (最小参数)
  python weapm_cli.py add-node --cluster-name LOG008 --address 127.0.0.2 --role write --cpulimit 8 --memlimit 16

  # 添加集群节点 (完整参数)
  python weapm_cli.py add-node --cluster-name LOG008 --address 127.0.0.3 --role master \\
    --cpulimit 16 --memlimit 32 --topic log_topic --bucketnames log_bucket

  # 删除集群节点
  python weapm_cli.py delete-node --ip 192.168.1.100

  # 使用自定义 API 地址
  python weapm_cli.py --base-url http://192.168.1.100:8080 dashboard

  # 指定环境
  python weapm_cli.py --env prod clusters
        """
    )

    # 全局参数
    parser.add_argument('--config', '-c', help='配置文件路径')
    parser.add_argument('--env', '-e', choices=['dev', 'prod'], help='环境名称 (dev/prod)')
    parser.add_argument('--base-url', help='API 基础 URL')
    parser.add_argument('--username', help='用户名')
    parser.add_argument('--password', help='密码')
    parser.add_argument('--timeout', type=int, default=30, help='请求超时时间(秒)')
    parser.add_argument('--quiet', '-q', action='store_true', help='静默模式,不输出日志')
    parser.add_argument('--output', '-o', choices=['json', 'pretty'], default='pretty',
                       help='输出格式 (默认: pretty)')

    subparsers = parser.add_subparsers(dest='command', help='可用命令')

    # dashboard 命令
    dashboard_parser = subparsers.add_parser('dashboard', help='获取数据大盘信息')

    # clusters 命令
    clusters_parser = subparsers.add_parser('clusters', help='集群管理')
    clusters_parser.add_argument('--detail', '-d', action='store_true', help='显示集群详情')
    clusters_parser.add_argument('--cluster-name', '-n', help='集群名称')

    # subsystems 命令
    subsystems_parser = subparsers.add_parser('subsystems', help='子系统管理')
    subsystems_parser.add_argument('--search', '-s', action='store_true', help='搜索子系统')
    subsystems_parser.add_argument('--subsys-id', help='子系统ID (用于搜索)')
    subsystems_parser.add_argument('--check', '-c', help='检查子系统是否存在')
    subsystems_parser.add_argument('--detail', '-d', help='获取子系统详情')
    subsystems_parser.add_argument('--limit', '-l', type=int, default=20, help='返回结果数量限制')

    # add-node 命令
    add_node_parser = subparsers.add_parser('add-node', help='添加集群节点')
    add_node_parser.add_argument('--cluster-name', required=True, help='集群名称')
    add_node_parser.add_argument('--address', required=True, help='节点IP地址')
    add_node_parser.add_argument('--role', required=True, help='节点角色')
    add_node_parser.add_argument('--cpulimit', help='CPU限制')
    add_node_parser.add_argument('--memlimit', help='内存限制')
    add_node_parser.add_argument('--topic', help='Topic')
    add_node_parser.add_argument('--bucketnames', help='存储桶名称')
    add_node_parser.add_argument('--backenddomain', help='后端域')
    add_node_parser.add_argument('--storagedomain', help='存储域')
    add_node_parser.add_argument('--status', help='状态')

    # delete-node 命令
    delete_node_parser = subparsers.add_parser('delete-node', help='删除集群节点')
    delete_node_parser.add_argument('--ip', required=True, help='节点IP地址')

    return parser.parse_args()


# ==================== 主函数 ====================

def main():
    """主函数 - 命令行入口"""
    args = parse_args()

    # 如果没有指定命令,显示帮助
    if not args.command:
        print("请指定命令。使用 --help 查看帮助信息。")
        sys.exit(1)

    # 配置日志级别
    if args.quiet:
        logging.getLogger().setLevel(logging.ERROR)

    # 加载配置
    try:
        if args.config or args.env:
            config = WeapmConfig.from_yaml(args.config, args.env)
        elif args.base_url:
            # 使用命令行参数创建配置
            config = WeapmConfig(
                base_url=args.base_url,
                username=args.username or "weapmUser",
                password=args.password or "Weapm@123admin",
                timeout=args.timeout
            )
        else:
            # 默认使用配置文件
            config = WeapmConfig.from_yaml()
    except FileNotFoundError as e:
        print(f"⚠️  {e}")
        print("请先创建配置文件 config.yaml,参考 config.yaml.example")
        sys.exit(1)
    except ValueError as e:
        print(f"⚠️  配置错误: {e}")
        sys.exit(1)

    # 创建客户端
    with WeapmClient(config) as client:
        try:
            # 根据命令执行相应操作
            if args.command == 'dashboard':
                cmd_dashboard(args, client)
            elif args.command == 'clusters':
                cmd_clusters(args, client)
            elif args.command == 'subsystems':
                cmd_subsystems(args, client)
            elif args.command == 'add-node':
                cmd_add_node(args, client)
            elif args.command == 'delete-node':
                cmd_delete_node(args, client)
            else:
                print(f"未知命令: {args.command}")
                sys.exit(1)

        except Exception as e:
            print(f"❌ 错误: {str(e)}")
            sys.exit(1)


if __name__ == "__main__":
    main()
